Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch

```diff
@@ -0,0 +1,956 @@
++        this.forceWrite = forceWrite;
++    // Paper end
+         this.c = immutableset;
+         this.d = flag;
+         this.g = datafixer;
+@@ -139,7 +146,7 @@ public class WorldUpgrader {
+                                     nbttagcompound2.remove("isLightOn");
+                                 }
+ 
+-                                if (flag2) {
++                                if (flag2 || forceWrite) { // Paper - add forceWrite flag
+                                     ichunkloader.a(chunkcoordintpair, nbttagcompound1);
+                                     flag1 = true;
+                                 }
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index dc7de2b59ec5ca3e5fba34dbb2aa2e6aed8f95cb..a6504a20f0bace4bb2e1397d44bf39d1a9b57776 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -1132,7 +1132,7 @@ public final class CraftServer implements Server {
+                 return true;
+             }, worlddata.getGeneratorSettings().d().d().stream().map((entry) -> {
+                 return ResourceKey.a(IRegistry.K, ((ResourceKey) entry.getKey()).a());
+-            }).collect(ImmutableSet.toImmutableSet()));
++            }).collect(ImmutableSet.toImmutableSet()), console.options.has("forceWrite")); // Paper - add forceWrite flag
+         }
+ 
+         long j = BiomeManager.a(creator.seed());
+@@ -2417,4 +2417,15 @@ public final class CraftServer implements Server {
+         return mobGoals;
+     }
+     // Paper end
++
++    // Paper start - API for compression settings
++    @Override
++    public io.papermc.paper.region.RegionCompressionType getRegionCompressionType() {
++        return com.destroystokyo.paper.PaperConfig.regionCompressionMethod;
++    }
++
++    public short getZstdDictVersion() {
++        return net.minecraft.server.RegionFileCompression.CURRENT_ZSTD_VERSION;
++    }
++    // Paper end
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
+index 010f702ea44d2146b0745b2b4d21f948d16cc424..cef1c9a9cb07ca4c52dbd13ba0dcf08e025a35ef 100644
+--- a/src/main/java/org/bukkit/craftbukkit/Main.java
++++ b/src/main/java/org/bukkit/craftbukkit/Main.java
+@@ -112,6 +112,7 @@ public class Main {
+                         .describedAs("Yml file");
+ 
+                 acceptsAll(asList("forceUpgrade"), "Whether to force a world upgrade");
++                acceptsAll(asList("forceWrite"), "Whether to force flushing to disk when upgrading"); // Paper - add forceWrite flag
+                 acceptsAll(asList("eraseCache"), "Whether to force cache erase during world upgrade");
+                 acceptsAll(asList("nogui"), "Disables the graphical console");
+ 
+diff --git a/src/main/resources/paper.zstd.dict.0 b/src/main/resources/paper.zstd.dict.0
```

Comment:
Where/how  was this found/generated?

Reply:
Using this project/technique: https://github.com/egg82/CompressionTests#zstd-dictionary
It's not incredibly well-written but it does what it's supposed to do and does it efficiently.

Appending `--dump` will create the raw chunk data for Zstd training, after using Chunky (or similar) to generate chunks.

Reply:
The exact commands I used, for reference:
```Bash
java -jar -Xmx8192m -Xms8192m compression-tests-1.1.2.jar ~/Servers/1.16.4/world/region --dump
find ~/Servers/1.16.4/world/region/rawchunks -type f -print0 | sort -zR | tail -zn +25001 | xargs -0 rm
zstd --train-fastcover ~/Servers/1.16.4/world/region/rawchunks/* -o paper.zstd.dict
```